src/workerd/api/deferred-proxy.h

```diff
@@ -0,0 +1,258 @@
+
+    result.value = DeferredProxy<T> { inner.get_return_object() };
+    return kj::_::PromiseNode::to<kj::Promise<DeferredProxy<T>>>(kj::_::OwnPromiseNode(this));
+  }
+
+  auto initial_suspend() { return inner.initial_suspend(); }
+  auto final_suspend() noexcept { return inner.final_suspend(); }
+  // Just trivially forward these.
+
+  void unhandled_exception() {
+    // Reject our outer promise if it hasn't yet been fulfilled, then forward to the inner
+    // implementation.
+
+    rejectOuterPromise();
+    inner.unhandled_exception();
+  }
+
+  kj::_::stdcoro::suspend_never yield_value(decltype(BEGIN_DEFERRED_PROXYING)) {
+    // This allows us to write `KJ_CO_MAGIC` within a DeferredProxyPromise<T> coroutine to fulfill
+    // the coroutine's outer promise with a DeferredProxy<T>.
+    //
+    // This could alternatively be implemented as an await_transform() with a magic parameter type.
+
+    fulfillOuterPromise();
+    return {};
+  }
+
+  template <CoroutineYieldValue<InnerCoroutineAdapter> U>
+  auto yield_value(U&& value) {
+    // Forward all other `co_yield`s to the inner coroutine, if it has a `yield_value()`
+    // implementation -- it might implement some magic, too.
+    return inner.yield_value(kj::fwd<U>(value));
+  }
+
+  void fulfill(kj::_::FixVoid<T>&& value) {
+    // Required by CoroutineMixin implementation to implement `co_return`.
+
+    fulfillOuterPromise();
+    inner.fulfill(kj::mv(value));
+  }
+
+  template <typename U>
+  auto await_transform(U&& awaitable) {
+    // Trivially forward everything, so we can await anything a kj::Promise<T> can.
+    return inner.await_transform(kj::fwd<U>(awaitable));
+  }
+
+  operator kj::_::CoroutineBase&() { return inner; }
+  // Required by Awaiter<T>::await_suspend() to support awaiting Promises.
+
+private:
+  void fulfillOuterPromise() {
+    // Fulfill the outer promise if it hasn't already settled.
+
+    if (!deferredProxyingHasBegun) {
+      // Our `result` is put in place already by `get_return_object()`, so all we have to do is arm
+      // the event.
+      onReadyEvent.arm();
+      deferredProxyingHasBegun = true;
+    }
+  }
+
+  void rejectOuterPromise() {
+    // Reject the outer promise if it hasn't already settled.
+
+    if (!deferredProxyingHasBegun) {
+      result.addException(kj::getCaughtExceptionAsKj());
+      onReadyEvent.arm();
+      deferredProxyingHasBegun = true;
+    }
+  }
+
+  // PromiseNode implementation
+
+  void destroy() override {
+    // The promise returned by `inner.get_return_object()` is what actually owns this coroutine
+    // frame. We temporarily store that in `result` until our outer promise is fulfilled. So, to
+    // destroy ourselves, we must manually drop `result`.
+    //
+    // On the other hand, if our outer promise has already been fulfilled, and `result` delivered to
+    // wherever it is going, then someone else directly owns the coroutine now, not us, and it's
+    // okay for this drop to be a no-op.
```

Comment:
I'm struggling to understand this. How does this object get destroyed if destroy() is a no-op?

Have you run your test under a memory leak checker?

Reply:
I guess the answer is that `inner` holds onto the outer coroutine handle and so inner's `destroy()` actually calls `handle.destroy()` which destroys the outer coroutine... this is very mind-bendy.

Reply:
What happens if `destroy()` ends up being called on the outer PromiseNode _after_ the inner one is destroyed? Then I guess this method is called on an already-freed object (probably segfaulting on the vtable). Can we definitively say this can't happen in practice?

Reply:
> Have you run your test under a memory leak checker?

Valgrind reports no leaks.

> this is very mind-bendy

Yep.

> What happens if destroy() ends up being called on the outer PromiseNode after the inner one is destroyed? Then I guess this method is called on an already-freed object (probably segfaulting on the vtable). Can we definitively say this can't happen in practice?

That's a great point -- incorrect destruction order would cause a segfault, so the second paragraph of this comment that I wrote is BS.

I think we _can_ definitively say that this won't happen in practice, because if kj-async failed to destroy PromiseNodes immediately after calling `get()`, then we'd have other problems. For example, some `.then()` continuations may rely on `kj::defer()`s running inside of attachments to their dependency promises before the continuations are invoked -- if those dependencies were not destroyed, then that guarantee would no longer hold. If you grep through async.c++, you'll see that every call to `get()` is immediately succeeded by a nullification of the node whose `get()` was called. (Except ironically, the one that backs `co_await`. That one is safe because the PromiseNode is in a temporary value that is destroyed after the `co_await expression is evaluated, though perhaps we should make that explicit.)

It'd be nice to have a stronger guarantee, though, so I pushed a commit which adds a test which forces a `get()` without immediate dependency destruction, and adds a mechanism to destroy the outer PromiseNode in its own `get()` via its self-pointer. Let me know what you think.

Aside: it has occurred to me that perhaps `get()` should somehow imply `destroy()`, and that `onReady()` should somehow imply `setSelfPointer()`...

Reply:
Hmm, I think I'm happy with your argument for why this was already a safe assumption and it's not necessary to add complexity to guarantee it. I'll approve this and leave it up to you if you want to revert that last commit (maybe keeping the test?).

Meanwhile, though, I do like the idea of `get()` implying `destroy()`. This might be a nice optimization for KJ promises across the board, avoiding a redundant virtual call. (But that would be a separate PR...)

Reply:
The Mac build seemed to fail spuriously, so I force-pushed a rebase on `main`. The HEAD commit of this PR is still identical to the previous 82d2aff.

Reply:
> Hmm, I think I'm happy with your argument for why this was already a safe assumption and it's not necessary to add complexity to guarantee it. 

I feel the same ... but I think I'd like to keep the safeguard, at least until `get()` can imply `destroy()`. The reason is that I have personally introduced exactly such incorrect destruction order bugs in kj-async while iterating on unrelated async PRs in the now-distant past, and it'd be nice if they were very noisy failures. The new test case I introduced didn't actually fail until I ran it under Valgrind, underscoring how insidious memory access problems are. :/

> I'll approve this and leave it up to you if you want to revert that last commit (maybe keeping the test?).

We couldn't actually keep the test if we remove the safeguard, since the test intentionally misuses the private `kj::_` API to trigger the problem.